src/torrent.cpp

```diff
@@ -2792,6 +2792,11 @@ namespace libtorrent {
 			TORRENT_ASSERT(valid_endpoints <= ae.endpoints.size());
 			ae.endpoints.erase(ae.endpoints.begin() + int(valid_endpoints), ae.endpoints.end());
 
+			// even if we're not listening on any sockets, we want to announce to
+			// trackers
+			if (ae.endpoints.empty())
```

Comment:
is there a legitimate case of when all endpoints get erased? should this have a different condition?

Reply:
`ae.endpoints` will be empty if and only if there are no applicable listen interfaces, so this is the correct condition. It means either there are no listen interfaces, or this is an ssl torrent and there are no ssl listen interfaces or vice-versa.

Reply:
actually, I just realized that if there's an error connecting to a tracker, the endpoint is removed and ``announce_with_tracker()`` is called again. So the retry logic relies on an empty list here being the "base-case". I detected this while working on the udp test, where a failure effectively caused an infinite loop (of immediately retrying the announce over and over again)

Reply:
I think the way this used to work was by recording the fail count in the ``announce_entry`` itself. Now it's stored in the announce_endpoint instead, and is lost (on line 2793). once it's erased, there's no trace that it failed, it will be added back and fail instantly again

Reply:
That sounds right. We probably should not remove the "no listen socket" endpoint in the code above.

Reply:
another approach would be to have a fake listen socket in the session.

Reply:
Then you'd have to worry about handling the not-really-a-listen-socket everywhere else that deals with listen sockets. Seems like it could be a bit of a land mine.

Reply:
If we're aiming to enable tracker announces _and_ keep the DHT going without a configured listen socket then it probably does make sense to create a "default" listen socket which actually does have a UDP socket but no TCP socket. The code should be able to handle this case anyways if it doesn't already. The only special case would be to hide the default socket from the UTP connection manager.

Reply:
yeah, I'm leaning towards doing that. At least for the UDP case, a lot of places assume a ``listen_socket_t``. I could just add a flag to say we should accept incoming connections on it (which applies both for uTP and TCP).